Alerting: refactor scheduler and separate notification logic

yuri-tceretian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Scheduler service is overloaded with many responsibilities: 
- loading changes from database: alert rules, the external sender
- providing external alertmanager URLs to API
- evaluating alert rules, 
- updating the database with evaluation states 
This makes scheduler hard to test and read.

This PR moves all logic related to sending notifications to a different service that is run independently and is consumed by the scheduler. The new service is in package `sender` and is called `AlertsRouter`. 

**Special notes for your reviewer**:
Note that this refactoring is not final. I did not change the schedulers test to make sure that nothing is broken as well as reduce the scope of the changes. In a follow-up PR I will update and move some of them from the scheduler package to the sender.

yuri-tceretian committed:
move registry to its own file

yuri-tceretian committed:
move state management from compat to scheduler

yuri-tceretian committed:
remove unused argument

yuri-tceretian committed:
extract notification logic to sender Dispatcher

ryantxu commented:
I am getting a panic in main now...  
```
panic: duplicate metrics collector registration attempted

goroutine 558 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0x104539edb?, {0x14000fece30?, 0x1, 0x8?})
	/Users/ryan/go/pkg/mod/github.com/prometheus/client_golang@v1.12.1/prometheus/registry.go:403 +0x88
github.com/prometheus/client_golang/prometheus/promauto.Factory.NewGauge({{0x105678468?, 0x1400018f630?}}, {{0x104539edb, 0x7}, {0x10453c66e, 0x8}, {0x1045d03f0, 0x2b}, {0x1045dc6a5, 0x2f}, ...})
	/Users/ryan/go/pkg/mod/github.com/prometheus/client_golang@v1.12.1/prometheus/promauto/auto.go:297 +0xf8
github.com/grafana/grafana/pkg/services/alerting/metrics.NewTickerMetrics({0x105678468, 0x1400018f630})
	/Users/ryan/workspace/grafana/grafana/pkg/services/alerting/metrics/metrics.go:16 +0x94
github.com/grafana/grafana/pkg/services/alerting.(*AlertEngine).Run(0x14000bb02a0, {0x10567bf58, 0x1400044ae00})
	/Users/ryan/workspace/grafana/grafana/pkg/services/alerting/engine.go:96 +0x4c
github.com/grafana/grafana/pkg/server.(*Server).Run.func1()
	/Users/ryan/workspace/grafana/grafana/pkg/server/server.go:157 +0x11c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/Users/ryan/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
	/Users/ryan/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x90
```

I am able to work around it with:
<img width="1031" alt="image" src="https://user-images.githubusercontent.com/705951/164815963-8b8b2d8e-4a87-4e5f-a47c-5ff96d70a392.png">

grobinson-grafana commented on pkg/services/ngalert/schedule/registry.go:
```diff
@@ -0,0 +1,156 @@
+
+	f := func() {
+		r.mu.Lock()
+		defer r.mu.Unlock()
```
I think this can deadlock if the consumer doesn't read the channel (i.e. it receives cancelation). It will also block all other goroutines if the consumer takes a long time to iterate all of the items.

yuri-tceretian replied:
I copied the code as-is from the scehdule.go.

joeblubaugh replied:
Then let's file a bug somewhere, too.

yuri-tceretian replied:
> I think this can deadlock if the consumer doesn't read the channel (i.e. it receives cancelation). It will also block all other goroutines if the consumer takes a long time to iterate all of the items.

I took a look at this closer, and although I do not understand why such a simple operation needs to be that complex, I do not see a case when the situation can happen. The method is used only in one place, a few lines below, the channel is read until it's closed. I would appreciate some examples of when it can fail.

grobinson-grafana commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -549,9 +347,21 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key models.AlertRul
+		toNotify := make([]*state.State, 0, len(processedStates))
+		sent := sch.clock.Now()
+		for _, alertState := range toNotify {
+			if !alertState.NeedsSending(sch.stateManager.ResendDelay) {
```
I think this should be in the notifier rather than the scheduler?

yuri-tceretian replied:
I think the notifier should not decide whether a state needs to be sent or not. I think ideally, that should be decided by the state manger. However, as long as this is refactoring, I try to limit scope of the changes.

grobinson-grafana replied:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent, but the state manager is responsible to deciding if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

grobinson-grafana replied:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent and the state manager should decide if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

grobinson-grafana replied:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent and the state manager should decide if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

grobinson-grafana replied:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent and the state manager should decide if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

grobinson-grafana replied:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent and the state manager should decide if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

grobinson-grafana commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -549,9 +347,21 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key models.AlertRul
+				continue
+			}
+			toNotify = append(toNotify, alertState)
+			alertState.LastSentAt = sent
```
I know its a `*.state.State` but I think it would avoid confusion if we set `LastSentAt` before `append`?

yuri-tceretian commented:
@ryantxu yea, sorry for that. The bug was fixed in https://github.com/grafana/grafana/pull/48190

yuri-tceretian commented on pkg/services/ngalert/schedule/registry.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func (r *alertRuleRegistry) iter() <-chan models.AlertRuleKey {
+	c := make(chan models.AlertRuleKey)
```
good suggestion but it think it is worth addressing in a follow-up PR. This PR is a refactoring and therefore I try not to change logic

yuri-tceretian commented on pkg/services/ngalert/schedule/compat.go:
```diff
@@ -109,21 +108,12 @@ func errorAlert(labels, annotations data.Labels, alertState *state.State, urlStr
 }
 
-func FromAlertStateToPostableAlerts(firingStates []*state.State, stateManager *state.Manager, appURL *url.URL) apimodels.PostableAlerts {
+func FromAlertStateToPostableAlerts(firingStates []*state.State, appURL *url.URL) apimodels.PostableAlerts {
```
there are some tests that rely on this method for now.. We will address it in a follow up PR.

yuri-tceretian replied:
done

yuri-tceretian committed:
move compat to sender

yuri-tceretian committed:
create constructor

yuri-tceretian committed:
add notifier dispatcher to API

yuri-tceretian committed:
make dispatcher's fields public

temporary to not refactor tests

yuri-tceretian committed:
update configuration of ngalert and scheduler

yuri-tceretian committed:
update tests to use dispatcher

yuri-tceretian committed:
make compat methods package private

yuri-tceretian committed:
service description

yuri-tceretian committed:
fixup

papagian commented on pkg/services/ngalert/schedule/alertnotifier_mock.go:
```diff
@@ -0,0 +1,98 @@
+
+// Expire provides a mock function with given fields: key, states
+func (_m *FakeAlertNotifier) Expire(key models.AlertRuleKey, states []*state.State) error {
+	ret := _m.Called(key, states)
```
CI tests seem to fail here.

joeblubaugh commented on pkg/services/ngalert/api/api.go:
```diff
@@ -60,26 +61,27 @@ type AlertingStore interface {
+	Policies               *provisioning.NotificationPolicyService
+	ContactPointService    *provisioning.ContactPointService
+	Templates              *provisioning.TemplateService
+	NotificationDispatcher *sender.Dispatcher
```
The alertmanager model calls these "Alerts" or "Alert Instances" - maybe `AlertDispatcher`? I think it would be good to make the meaning of "Notification" be "Something sent from Grafana to a Contact Point"

yuri-tceretian replied:
Good Idea

joeblubaugh commented on pkg/services/ngalert/api/api_admin.go:
```diff
@@ -16,7 +16,7 @@ import (
 
 type AdminSrv struct {
-	scheduler Scheduler
+	scheduler ExternalAlertmanagerProvider
```
Should this still be called the scheduler when what it does is provide Alertmanagers?

yuri-tceretian replied:
1. It doesn't anymore. Dispatcher does
2. The interface is declared on the user side which probably don't care about Dispatcher and cares only about some of its methods

yuri-tceretian replied:
Renamed

joeblubaugh commented on pkg/services/ngalert/ngalert.go:
```diff
@@ -81,8 +83,9 @@ type AlertNG struct {
-	MultiOrgAlertmanager *notifier.MultiOrgAlertmanager
-	accesscontrol        accesscontrol.AccessControl
+	MultiOrgAlertmanager   *notifier.MultiOrgAlertmanager
+	NotificationDispatcher *sender.Dispatcher
```
Same ? as above

joeblubaugh commented on pkg/services/ngalert/ngalert.go:
```diff
@@ -106,31 +109,38 @@ func (ng *AlertNG) init() error {
-		AdminConfigPollInterval: ng.Cfg.UnifiedAlerting.AdminConfigPollInterval,
-		DisabledOrgs:            ng.Cfg.UnifiedAlerting.DisabledOrgs,
-		MinRuleInterval:         ng.Cfg.UnifiedAlerting.MinInterval,
+		return fmt.Errorf("failed to initialize alerting because notifier manager fail to warm up: %w", err)
```
What does "notifier manager" mean? It's the alertmanager startup that fails here.

yuri-tceretian replied:
renamed

joeblubaugh commented on pkg/services/ngalert/ngalert.go:
```diff
@@ -106,31 +109,38 @@ func (ng *AlertNG) init() error {
 		appUrl = nil
 	}
+
+	clk := clock.New()
```
Do we use any other `clock` package APIs in this function? Seems fine to shadow the package name here instead of ambiguity about whether is is a "clock" or a "click" - I initially read "click"

joeblubaugh commented on pkg/services/ngalert/sender/dispatcher.go:
```diff
@@ -0,0 +1,280 @@
+package sender
```
What's new in this file, and what is copied from other code? I want to know how closely to review code.

yuri-tceretian replied:
Almost 100% copy/past

joeblubaugh commented on pkg/services/ngalert/sender/dispatcher.go:
```diff
@@ -0,0 +1,280 @@
+		appURL:                  appURL,
+		disabledOrgs:            disabledOrgs,
+		adminConfigPollInterval: configPollInterval,
+	}
```
Do you want to do an initial `SyncAndApplyConfigFromDatabase` before returning? 

edit: I'm leaving this comment here because I think it illustrates my point about requiring the user to call SyncAndApply themselves being confusing.

joeblubaugh commented on pkg/services/ngalert/sender/dispatcher.go:
```diff
@@ -0,0 +1,280 @@
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+)
+
+// Dispatcher is a service that handles alert notifications. It accepts an alert that is represented by models.AlertRuleKey and state.State,
```
Some suggested edits, I welcome your feedback on these:
```go
// Dispatcher handles alerts generated during alert rule evaluation. It
// converts alerts represeneted by {models.AlertRuleKey, state.State}
// into the standard Prometheus Alertmanager format, and then sends
// the resulting alert to all configured alertmanagers.
//
// After creating a Dispatcher, you must call Run to keep the Dispatcher's
// state synchronized with the alerting configuration.
```

joeblubaugh commented on pkg/services/ngalert/sender/dispatcher.go:
```diff
@@ -0,0 +1,280 @@
+	return d
+}
+
+func (d *Dispatcher) adminConfigSync(ctx context.Context) error {
```
If you don't want to synchronize in the constructor, you might want to lead this function off with a call to `SyncAndApplyConfigFromDatabase`. Otherwise there will be an initial period of 60s with no configuration on the Dispatcher, right?

edit: I'm leaving this comment here because I think it illustrates my point about requiring the user to call SyncAndApply themselves being confusing.

joeblubaugh commented on pkg/services/ngalert/ngalert.go:
```diff
@@ -106,31 +109,38 @@ func (ng *AlertNG) init() error {
+
+	dispatcher := sender.NewDispatcher(ng.MultiOrgAlertmanager, store, clk, appUrl, ng.Cfg.UnifiedAlerting.DisabledOrgs, ng.Cfg.UnifiedAlerting.AdminConfigPollInterval)
+
+	if err := dispatcher.SyncAndApplyConfigFromDatabase(); err != nil {
```
I wrote a bunch of comments in `sender.go` below about where to do this Sync before seeing that it's done here after construction and before calling Run. I think the API would be simpler and less error prone if we didn't require the programmer to make this additional call when we know what the right thing to do is.

yuri-tceretian replied:
After more thinking, I decided to address it in part-2 which will refactor tests. Currently, this small change could cause problems in scheduler's unit tests because they use the real dispatcher.

joeblubaugh replied:
Sounds good. I'm also fine with keeping the initial sync separate if that's consistent with how we configure all the services in this package.

joeblubaugh commented on pkg/services/ngalert/sender/dispatcher.go:
```diff
@@ -0,0 +1,280 @@
+	logger := d.logger.New("rule_uid", key.UID, "org", key.OrgID)
+	// Send alerts to local notifier if they need to be handled internally
+	// or if no external AMs have been discovered yet.
+	var localNotifierExist, externalNotifierExist bool
```
nit: `s/Exist/Exists`

yuri-tceretian replied:
this is moved as is.

yuri-tceretian committed:
rename structs and interfaces

yuri-tceretian committed:
better logs

yuri-tceretian committed:
better comment

yuri-tceretian committed:
fix tests

gotjosh reviewed (changes requested):
I haven't looked at this PR in detail (perhaps none at all) but as I mentioned before in: https://github.com/grafana/grafana/pull/40341#issuecomment-953117885

I'd advise against using the name dispatcher for the separate component. This would cause more confusion.

I apologise in advance, that even though I have requested changes for this, it's very unlikely that I'll get to review this PR for the next few weeks given my other priorities.

joeblubaugh commented:
Some other suggestions if `Dispatcher` or `AlertDispatcher` are too overloaded. These are all over the place.

The Prometheus core uses a component called: `notifier.Manager`, which performs very similar functions as the `sender.AlertDispatcher`. I don't like that name very much, but it's a possibility. We use the package `notifier` for the multiorg alertmanager. That package name could change to something like `alert` so we have `alert.Manager` and `alert.MultiOrgManager`, and we could then use the `notifier` package (or call it `notify`?) for this code.

`sender.Notifier`? The primary method is `Notify`. `sender.Notifier.Enqueue()`? `sender.AlertQueue`?

grobinson-grafana commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -49,6 +37,12 @@ type ScheduleService interface {
 }
 
+//go:generate mockery --name AlertSender --structname FakeAlertSender --inpackage --filename alertsender_mock.go --with-expecter
+type AlertSender interface {
```
I'm not sure this is the right interface as states are not alerts. What does it mean to send a state?

grobinson-grafana commented on pkg/services/ngalert/sender/compat.go:
```diff
@@ -109,33 +108,21 @@ func errorAlert(labels, annotations data.Labels, alertState *state.State, urlStr
-		alertState.LastSentAt = ts
-		sentAlerts = append(sentAlerts, alertState)
 	}
-	stateManager.Put(sentAlerts)
```
I'm pleased this line was removed, it didn't make sense to me either!

yuri-tceretian commented:
> I would like to see better separation between the two because a state is not an alert and I don't think it makes sense to send a state. It feels odd to me that we don't have a good separation between states and alerts.

we do, it's PostableAlerts and compat.go as a shim, I just moved it to sender. I can move it back so state would be converted to PostableAlerts and submitted to the sender.

yuri-tceretian committed:
rollback compat moving

yuri-tceretian committed:
rename interface to AlertsSender

yuri-tceretian committed:
rename Dispatcher to AlertsRouter

yuri-tceretian committed:
revert split of registry from schedule

yuri-tceretian commented:
I reverted some changes that are not directly related to the split of the alert router and scheduler to reduce the scope of changes and increase the chances of this change being merged. 
Also, I moved back compat.go that converts alert state to PostableAlert. As @grobinson-grafana pointed out the interface that accepted state looked confusing.

torkelo commented:
backported to v9 or moved to v9.1 ?   , trying to cleanup v9 milestone

yuri-tceretian committed:
Merge branch 'up/main' into scheduler-dispatcher

# Conflicts:
#	pkg/services/ngalert/api/api.go
#	pkg/services/ngalert/ngalert.go
#	pkg/services/ngalert/schedule/schedule_test.go
#	pkg/services/ngalert/schedule/schedule_unit_test.go

yuri-tceretian committed:
Merge branch 'up/main' into scheduler-dispatcher

# Conflicts:
#	pkg/services/ngalert/api/api.go
#	pkg/services/ngalert/ngalert.go
#	pkg/services/ngalert/schedule/schedule.go
#	pkg/services/ngalert/schedule/schedule_unit_test.go

yuri-tceretian committed:
misc fixes

yuri-tceretian committed:
rename remaining dispatcher to router

yuri-tceretian committed:
remove tag

yuri-tceretian committed:
Merge branch 'up/main' into scheduler-dispatcher

# Conflicts:
#	pkg/services/ngalert/ngalert.go
#	pkg/services/ngalert/schedule/schedule.go
#	pkg/services/ngalert/schedule/schedule_unit_test.go

joeblubaugh reviewed (commented):
I definitely think we should get this merged - separating scheduling and sending made this whole codebase simpler for me to understand, and with a little more work the tests for each component can be separated as well.

Noted below: we should finish the separation by making the `Senders` and `AdminConfigMtx` unexported and separate testing to verify the router's config application separately from the scheduler's tests.

yuri-tceretian commented on pkg/services/ngalert/ngalert.go:
```diff
@@ -202,14 +214,18 @@ func (ng *AlertNG) Run(ctx context.Context) error {
 
 	children, subCtx := errgroup.WithContext(ctx)
 
+	children.Go(func() error {
```
technically, no. I just grouped them together

yuri-tceretian commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -55,6 +44,12 @@ type ScheduleService interface {
 	folderUpdateHandler(ctx context.Context, evt *events.FolderUpdated) error
 }
 
+//go:generate mockery --name AlertsSender --structname FakeAlertsSender --inpackage --filename alerts_sender_mock.go --with-expecter
```
+1 Great suggestion!

yuri-tceretian committed:
better log messages

yuri-tceretian committed:
rename fake to mock

yuri-tceretian committed:
simplify scheduler.Run

yuri-tceretian commented on pkg/services/ngalert/schedule/schedule_unit_test.go:
```diff
@@ -53,19 +53,19 @@ func TestSendingToExternalAlertmanager(t *testing.T) {
 	// Make sure we sync the configuration at least once before the evaluation happens to guarantee the sender is running
 	// when the first alert triggers.
-	require.NoError(t, sched.SyncAndApplyConfigFromDatabase())
-	sched.adminConfigMtx.Lock()
```
100% agree. I did not intend to split tests in this PR but follow-up in another PR after we merge it.

yuri-tceretian committed:
rename fake to mock in tests

joeblubaugh reviewed (approved):
LGTM! @gotjosh can you confirm whether your requested changes have been addressed?

gotjosh commented on pkg/services/ngalert/sender/router.go:
```diff
@@ -0,0 +1,268 @@
+	return s.DroppedAlertmanagers()
+}
+
+// Run starts regular updates of the configuration
```
```suggestion
// Run starts regular updates of the configuration.
```

gotjosh commented on pkg/services/ngalert/sender/router.go:
```diff
@@ -0,0 +1,268 @@
+	return d
+}
+
+func (d *AlertsRouter) adminConfigSync(ctx context.Context) error {
```
I think we should inline this within the `run` function - I don't see us gaining anything from this level of indirection. It doesn't help with testing and it's [quite](https://github.com/prometheus/prometheus/blob/main/rules/manager.go#L341-L350) [common](https://github.com/grafana/mimir/blob/main/pkg/ruler/ruler.go#L428-L458) for control functions (e.g. `run`) to have this structure.

yuri-tceretian replied:
Agree. Moved it.

gotjosh commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -532,55 +346,14 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
-		notify(expiredAlerts, logger)
+		err := sch.notifier.Send(key, expiredAlerts)
+		if err != nil {
+			logger.Error("failed to send alerts to notifier", "err", err)
```
Ditto from my comment re "notifier", "router" and "sender.

gotjosh commented on pkg/services/ngalert/sender/router.go:
```diff
@@ -0,0 +1,268 @@
+	}
+
+	if !localNotifierExist && !externalNotifierExist {
+		return fmt.Errorf("no external or internal notifier - [%d] alerts not delivered", len(alerts.PostableAlerts))
```
The caller is not doing anything with this error - so I see no reason to return it (I believe this is one of the functions introduced by this refactor so we can modify it). Can we just log the error here and not return anything from this function.

yuri-tceretian replied:
I would like to keep the error here because it is logged by the scheduler that have much more context in the logger (the tick, version of the alert rule, rule key

gotjosh replied:
Can you please show an except of how these log lines look like? I think we'll notice that we have enough information as it is from the logs that this is outputting e.g. on error you'll log twice from two different places.

yuri-tceretian replied:
I took a look again and it was just that single error that is returned. So, it does not make sense to handle just that one considering the fact that this is an edge case. I changed the interface

yuri-tceretian committed:
Apply suggestions from code review

Co-authored-by: gotjosh <josue.abreu@gmail.com>

yuri-tceretian committed:
Update pkg/services/ngalert/ngalert.go

Co-authored-by: gotjosh <josue.abreu@gmail.com>

yuri-tceretian commented on pkg/services/ngalert/ngalert.go:
```diff
@@ -151,7 +137,32 @@ func (ng *AlertNG) init() error {
+		Metrics:         ng.Metrics.GetSchedulerMetrics(),
+		DisabledOrgs:    ng.Cfg.UnifiedAlerting.DisabledOrgs,
+		MinRuleInterval: ng.Cfg.UnifiedAlerting.MinInterval,
+		AlertSender:     alertsRouter,
```
The AlertsSender interface is declared on scheduler side. I chose its name following the same conception like io.Reader and io.Writer interfaces. Agree that 'notifier' can be renamed to sender.

gotjosh replied:
I got tricked by the GitHub diff and it's not the `stateManager` is the `API` where it is called a router.

gotjosh replied:
> The AlertsSender interface is declared on scheduler side.

This is not true. Within `router.go`, I see `Senders map[int64]*Sender` so now we have:

`AlertsSender` in the schedule package
`AlertsRouter` that has multiple `Senders` in the `sender` package. 

In my eyes, if `router -> sender` are part of the internals of `sender` package and the scheduler _only_ interacts with `router` then it should be `router` *everywhere*.

What I want to avoid is the confusion that when we talk about the "alerts sender" we now need to ask - the sender in sender package or the sender in the schedule package?

yuri-tceretian replied:
yes, the `Senders` field is exported now but it will be hidden in the package later when unit tests are updated, and it as well as the struct `Sender` will not be exported anywhere. 
As an alternative, I can think of `notifier` however, it also conflicts with `notifier` package where it represents a `contact point`. I would rather rename `Sender` that is in the `sender` package as well as the package itself to reflect that this is a client for external Alertmanagers. I think it will make more sense if we merge `sender` and `notifier` packages under the same package, e.g. `notifications`. 
I also do not think that this is a good idea to name the interface AlertsRouter as it will interfere with the struct itself.

yuri-tceretian committed:
get rid of term notifier in schedule package

yuri-tceretian committed:
move mutex declaration close to what it protects

gotjosh commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -55,6 +43,12 @@ type ScheduleService interface {
 }
 
+//go:generate mockery --name AlertsSender --structname AlertsSenderMock --inpackage --filename alerts_sender_mock.go --with-expecter
+// AlertsSender is an interface for a service that is responsible for sending notifications to the end-user
```
```suggestion
// AlertsSender is an interface for a service that is responsible for sending notifications to the end-user.
```

gotjosh commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -640,8 +413,10 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
-		notify(alerts, logger)
+		err = sch.notifier.Send(key, alerts)
+		if err != nil {
+			logger.Error("failed to send alerts to notifier", "err", err)
```
Please see my comment above, I think we need a better term that's not sender with the scheduler package otherwise we have the same problem we had with the dispatcher. To me, router *everywhere* is a good compromise but if you don't like that I'd like to see another proposal that's not *sender*.

yuri-tceretian committed:
Merge branch 'up/main' into scheduler-dispatcher

# Conflicts:
#	pkg/services/ngalert/ngalert.go
#	pkg/services/ngalert/schedule/schedule.go
#	pkg/services/ngalert/schedule/schedule_test.go
#	pkg/services/ngalert/schedule/schedule_unit_test.go

yuri-tceretian committed:
add .

yuri-tceretian committed:
move adminConfigSync to Run

yuri-tceretian committed:
Send to not return error

grobinson-grafana commented on pkg/services/ngalert/sender/router.go:
```diff
@@ -0,0 +1,268 @@
+	s, ok := d.Senders[key.OrgID]
+	if ok && d.SendAlertsTo[key.OrgID] != models.InternalAlertmanager {
+		logger.Debug("sending alerts to external notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts)
+		s.SendAlerts(alerts)
```
I looked at the code too and unless I have also misunderstood I think Josh is correct?

joeblubaugh replied:
On a second read after Josh's comment, I agree. I got crosseyed somewhere, because I ended up reading this code:
https://github.com/prometheus/prometheus/blob/44fcf876caad9a5d28b92a728d2c98c34015d377/notifier/notifier.go#L324

gotjosh commented on pkg/services/ngalert/schedule/schedule.go:
```diff
@@ -83,27 +77,20 @@ type schedule struct {
-	adminConfigPollInterval time.Duration
-	disabledOrgs            map[int64]struct{}
-	minRuleInterval         time.Duration
+	alertsSender    AlertsSender
```
My last comment that echoes what I've said as part of https://github.com/grafana/grafana/pull/48144#discussion_r918245869 (and still remains unaddressed) the comment there seems to be orthogonal is that I still don't agree with the name `alertsSender` here.

This is the router and not the sender. The sender is an internal component that the router uses and therefore it's confusing.

A suggestion is to change it to router - you don't need the alerts prefix because it can be `sender.Router` and the interface itself can come from the other package if it helps simplify things.

If you don't like my suggestion then I'd hope to see it changed to something else that's not `alertsSender` for both the attribute and the interface because *we already have a sender* that's being used within the router.

gotjosh replied:
Discussed this offline with Yuri, it seems like the proposal here is that we'll change the name of the sender within the router to something different in the next PR - I'm OK with that as long as it doesn't remain in this state for too long.

gotjosh reviewed (approved):
LGTM but I hope to see my comment being addressed either as part of this PR or the next one.
